src/zope/publisher/browser.py

```diff
@@ -247,16 +248,17 @@ def _createResponse(self):
 
     def _decode(self, text):
         """Try to decode the text using one of the available charsets."""
-        # According to PEP-3333, in python-3, QUERY_STRING is a string,
-        # representing 'latin-1' encoded byte array. So, if we are in python-3
-        # context, encode text as 'latin-1' first, to try to decode
-        # resulting byte array using user-supplied charset.
-        if not isinstance(text, bytes):
-            text = text.encode('latin-1')
         if self.charsets is None:
             envadapter = IUserPreferredCharsets(self)
             self.charsets = envadapter.getPreferredCharsets() or ['utf-8']
             self.charsets = [c for c in self.charsets if c != '*']
+        if not PYTHON2 and self.charsets and self.charsets[0] == 'utf-8':
+            # optimization: we are trying to decode something cgi.FieldStorage already
+            # decoded for us, let's just return it rather than waste time decoding
+            return text
+        if not PYTHON2:
+            # undo what cgi.FieldStorage did and maintain backwards compat
+            text = text.encode('utf-8')
```

Comment:
This was a bit hard to me to follow, with the partially overlapping if conditions.  I would suggest

```python
if not PYTHON2:
    if self.charsets and self.charsets[0] == 'utf-8':
        # optimization: ...
        return text
    # undo what cgi.FieldStorage did....
    text = text.encode('utf-8')
```

Reply:
Also, I'm guessing this is lossy?  `cgi.FieldStorage()` defaults to `encoding='UTF-8', errors='replace'` and once you've replaced invalid UTF-8 characters, there's no way of getting them back.

We could pass `latin-1` to ZopeFieldStorage() and then do our own charset conversion here.

Reply:
Ah, I've also discovered cases where cgi.FieldStorage gives us `b''` as the value, when the POSTed form contains a blank file field (in a request generated by zope.testbrowser/WebTest).